Problem set 2: Aggregation, reshaping, and validation

Due: 11:59pm on Friday, September 13 by uploading to Brightspace

Your name:

1. For this problem set, use any of the following datasets:

* [NOAA](https://www.ncei.noaa.gov/products) weather and climate data
* [World Happiness Report](https://worldhappiness.report/ed/2021/#appendices-and-data)
* [NASA data](https://data.nasa.gov/)
* [US Census data](https://data.census.gov/)
* [WHO COVID data](https://covid19.who.int/)
* [Netflix data](https://www.kaggle.com/datasets/ashishgup/netflix-rotten-tomatoes-metacritic-imdb)
* [Nashville open data](https://data.nashville.gov/)
* [Yelp reviews data](https://www.yelp.com/dataset)

In [3]:
# Read in the data
import pandas as pd
df = pd.read_excel('C:\\Users\\Ruan Jingyu\\OneDrive - PennO365\\Desktop\\Academic Courses\\DS 5610-01 Exploratory Data Analysis (2025F)\\WHR21_Data_Mortality.xlsx')

# Do some investigative EDA to identify the structure of the data and its content. Then write a brief summary of the dataset in your own words.
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 166 entries, 0 to 165
Data columns (total 17 columns):
 #   Column                                                                       Non-Null Count  Dtype  
---  ------                                                                       --------------  -----  
 0   Country name                                                                 166 non-null    object 
 1   Population 2020                                                              164 non-null    float64
 2   Population 2019                                                              164 non-null    float64
 3   COVID-19 deaths per 100,000 population in 2020                               163 non-null    float64
 4   Median age                                                                   163 non-null    float64
 5   Island                                                                       166 non-null    int64  
 6   Index of exposure to COVID-19  infections 

In [10]:
display(df.sample(10))

Country name  Population 2020  Population 2019  \
25          Iran       83992953.0       82913906.0   
130        Malta         441539.0         502653.0   
136    Nicaragua        6624554.0        6545502.0   
138         Oman        5106622.0        4974986.0   
80       Albania        2877800.0        2854191.0   
89       Bolivia       11673029.0       11513100.0   
18        Poland       37846605.0       37970874.0   
22        Sweden       10099270.0       10285453.0   
35         Kenya       53771300.0       52573973.0   
132     Mongolia        3278292.0        3225167.0   

     COVID-19 deaths per 100,000 population in 2020  Median age  Island  \
25                                        65.747200   32.400002       0   
130                                       49.599201   42.400002       1   
136                                        2.490700   27.299999       0   
138                                       29.354000   30.700001       0   
80                                        41.038300   38.000000       0   
89                                        78.514297   25.400000       0   
18                                        75.446701   41.799999       0   
22                                        86.412201   41.000000       0   
35                                         3.105700   20.000000       0   
132                                        0.030500   28.600000       0   

     Index of exposure to COVID-19  infections in other countries as of March 31  \
25                                            1.706699                             
130                                           2.822389                             
136                                           0.860422                             
138                                           1.075521                             
80                                            3.058284                             
89                                            0.608560                             
18                                            3.198972                             
22                                            2.589931                             
35                                            0.781459                             
132                                           1.139612                             

     Log of average distance to SARS countries  WHO Western Pacific Region  \
25                                    8.809122                           0   
130                                   9.096670                           0   
136                                   9.527919                           0   
138                                   8.780088                           0   
80                                    9.035891                           0   
89                                    9.726268                           0   
18                                    8.978474                           0   
22                                    8.963017                           0   
35                                    9.118932                           0   
132                                   8.326016                           1   

     Female head of government  Index of institutional trust  \
25                           0                      0.367223   
130                          0                      0.402808   
136                          0                      0.280303   
138                          0                      0.579144   
80                           0                      0.150139   
89                           0                      0.101161   
18                           0                      0.303745   
22                           0                      0.450582   
35                           0                      0.331046   
132                          0                      0.102140   

     Gini coefficient of income  All-cause death count, 2017  \
25                       39.97

### Summary
*This dataset, from World Hapiness Report, contains information on 166 countries, including their demographics (such as population size and age distribution), pandemic data (COVID-19 deaths and excess mortality), structural features (island status, exposure to international spread), and social-economic factors (institutional trust, income inequality, female leadership) in 2020. Most data are well documented, while records of all-cause mortality from previous years are less complete.*


2. Aggregate and reshape

In [14]:
# Reshape the data in some way that changes the unit of analysis. Convert it into an aggregated summary table that provides new information from that table
group_by_island = df.groupby("Island").agg({
    "Population 2020": "mean",
    "Median age": "mean",
    "COVID-19 deaths per 100,000 population in 2020": "mean",
    "Excess deaths in 2020 per 100,000 population, relative to 2017-2019 average": "mean",
    "Index of institutional trust": "mean"
}).reset_index()

group_by_island
# Write 3-4 sentences summarizing what new insight we get from the aggregated table


Island  Population 2020  Median age  \
0       0     4.947931e+07   30.101418   
1       1     3.234825e+07   34.295455   

   COVID-19 deaths per 100,000 population in 2020  \
0                                       33.172511   
1                                       14.883223   

   Excess deaths in 2020 per 100,000 population, relative to 2017-2019 average  \
0                                         119.959358                             
1                                          43.188946                             

   Index of institutional trust  
0                      0.308743  
1                      0.369632

### Summary

*From the aggregated data, we see island nations tend to have smaller populations and  older median ages compared to non-island countries. They had lower COVID-19 deaths and excess mortality rates, potentially due to their geographic isolation. Island countries also show somewhat higher institutional trust.*


3. Validation and missingness


In [ ]:
# Choose one numeric column from either one of the original tables or your newly merged table. Convert a random 5% sample of values to NaN. Make sure you set a seed first so the dropped values are always the same
import numpy as np

np.random.seed(42)

df_num = df.copy()

random_indices = np.random.choice(df_num.index, size=int(0.05 * len(df_num)), replace=False)
df_num.loc[random_indices, "COVID-19 deaths per 100,000 population in 2020"] = np.nan

# Calculate the percent of missing values in each column

missing_percent = df_num.isnull().mean() * 100
print("Percent of missing values in each column:")
print(missing_percent)

# Find the mean of the numeric column

mean_value = df_num["COVID-19 deaths per 100,000 population in 2020"].mean()
print(f"Mean of 'COVID-19 deaths per 100,000 population in 2020': {mean_value}")

# Use listwise deletion and find the mean again

mean_value_listwise = df_num["COVID-19 deaths per 100,000 population in 2020"].dropna().mean()
print(f"Mean of 'COVID-19 deaths per 100,000 population in 2020' (listwise deletion): {mean_value_listwise}")

# Use mean imputation and find the mean again

df_num["COVID-19 deaths per 100,000 population in 2020"].fillna(mean_value, inplace=True)
mean_value_imputed = df_num["COVID-19 deaths per 100,000 population in 2020"].mean()
print(f"Mean of 'COVID-19 deaths per 100,000 population in 2020' (mean imputation): {mean_value_imputed}")

Percent of missing values in each column:
Country name                                                                    0.000000
Population 2020                                                                 1.204819
Population 2019                                                                 1.204819
COVID-19 deaths per 100,000 population in 2020                                  6.626506
Median age                                                                      1.807229
Island                                                                          0.000000
Index of exposure to COVID-19  infections in other countries as of March 31     1.204819
Log of average distance to SARS countries                                       1.204819
WHO Western Pacific Region                                                      0.000000
Female head of government                                                       0.000000
Index of institutional trust                                        

C:\temp\ipykernel_11856\693371499.py:29: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_num["COVID-19 deaths per 100,000 population in 2020"].fillna(mean_value, inplace=True)
C:\temp\ipykernel_11856\693371499.py:40: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when 

In [ ]:
# Use a more advanced form of imputation and find the mean again
df_num = df.copy()
random_indices = np.random.choice(df_num.index, size=int(0.05 * len(df_num)), replace=False)
df_num.loc[random_indices, "COVID-19 deaths per 100,000 population in 2020"] = np.nan

# Median imputation
median_value = df_num["COVID-19 deaths per 100,000 population in 2020"].median()
df_num_median = df_num.copy()
df_num_median["COVID-19 deaths per 100,000 population in 2020"].fillna(median_value, inplace=True)

mean_value_median = df_num_median["COVID-19 deaths per 100,000 population in 2020"].mean()
print(f"Mean of 'COVID-19 deaths per 100,000 population in 2020' (median imputation): {mean_value_median}")


Mean of 'COVID-19 deaths per 100,000 population in 2020' (median imputation): 30.288275399447297


C:\temp\ipykernel_11856\3306536941.py:9: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_num_median["COVID-19 deaths per 100,000 population in 2020"].fillna(median_value, inplace=True)


### Compare these 4 values and in 3-4 sentences decribe that it tells us about imputation and missing data

*When comparing the four approaches, the values are close but not equal. Listwise deletion reduces the sample size and can slightly bias the result if the missing values are not random. Mean imputation keeps the mean stable but reduces the variance. Median imputation is less sensitive to outliers.*

In [18]:
# Repeat the above steps with a factor/string column, this time doing mode imputation instead of mean
df_cat = df.copy()
np.random.seed(42)

random_indices = np.random.choice(df_cat.index, size=int(0.05 * len(df_cat)), replace=False)
df_cat.loc[random_indices, "Female head of government"] = np.nan

missing_percent_cat = df_cat.isnull().mean() * 100
print("Percent of missing values in each column:")
print(missing_percent_cat)

# Original mode
true_mode = df["Female head of government"].mode()[0]

# Listwise deletion
listwise_mode = df_cat["Female head of government"].dropna().mode()[0]

# Mode imputation
mode_value = df_cat["Female head of government"].mode()[0]
df_cat_mode = df_cat.copy()
df_cat_mode["Female head of government"].fillna(mode_value, inplace=True)
mode_imputed = df_cat_mode["Female head of government"].mode()[0]

print(f"Original mode: {true_mode}")
print(f"Mode after listwise deletion: {listwise_mode}")
print(f"Mode after mode imputation: {mode_imputed}")

Percent of missing values in each column:
Country name                                                                    0.000000
Population 2020                                                                 1.204819
Population 2019                                                                 1.204819
COVID-19 deaths per 100,000 population in 2020                                  1.807229
Median age                                                                      1.807229
Island                                                                          0.000000
Index of exposure to COVID-19  infections in other countries as of March 31     1.204819
Log of average distance to SARS countries                                       1.204819
WHO Western Pacific Region                                                      0.000000
Female head of government                                                       4.819277
Index of institutional trust                                        

C:\temp\ipykernel_11856\3507767073.py:21: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_cat_mode["Female head of government"].fillna(mode_value, inplace=True)
